core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -600,6 +600,17 @@ private[spark] class AppStatusListener(
     liveUpdate(task, now)
 
     Option(liveStages.get((event.stageId, event.stageAttemptId))).foreach { stage =>
+      if (event.taskInfo.speculative) {
+        stage.speculationStageSummary = Some(stage.speculationStageSummary
+          .getOrElse(new LiveSpeculationStageSummary(event.stageId, event.stageAttemptId)))
+        val speculationSummary = stage.speculationStageSummary.get
+        speculationSummary.numActiveTasks += 1
+        speculationSummary.numTasks += 1
+      }
+      if(stage.speculationStageSummary.isDefined) {
+        maybeUpdate(stage.speculationStageSummary.get, now)
```

Comment:
With the latest change, `stage.speculationStageSummary` never become `None` once it's substituted `Some`. So, unnecessary updates may happen line you were afraid.
https://github.com/apache/spark/pull/34607#discussion_r753559069

Let's change `maybeUpdate` to return whether it actually updates or not.
Then, we can set `None` to `stage.speculationStageSummary` if the update actually happens.

Reply:
How about we make a deterministic update if the taskInfo is speculative and then we avoid `maybeUpdate` altogether ?

Reply:
By setting it to `None`, we will be losing on accumulated data of speculation summary

Reply:
```      if (event.taskInfo.speculative) {
        stage.speculationStageSummary = Some(stage.speculationStageSummary
          .getOrElse(new LiveSpeculationStageSummary(event.stageId, event.stageAttemptId)))
        val speculationSummary = stage.speculationStageSummary.get
        speculationSummary.numActiveTasks += 1
        speculationSummary.numTasks += 1
       update(stage.speculationSummary, now)
      }```
This means we only update when the task is speculative. It will also leads to a minimum number of updates

Reply:
> By setting it to None, we will be losing on accumulated data of speculation summary

Setting `None` means, currently updated speculation info is flushed.
Why updates will be lost?

Reply:
Since we are indexing by stageId and stageAttemptId, wont the second call to update overwrite the first ?
For example,

```
speculationSummary.numActiveTasks +=1
update(speculationSummary, now)

speculationSummary = new LiveStageSpeculationSummary(id, attempt)
update(speculationSummary, now)
```
The second update will overwrite the first one, right ? And for the stage, the speculation summary will show 0 tasks as the default value

Reply:
Ah, OK, `stage.speculationStageSummary` hold just a delta.
So how about:

* Revert he declaration of `stage.speculationStageSummary` (`val speculationStageSummary: LiveSpeculationStageSummary = ...`)
* Declare a variable flag like `needUpdateSpeculationSummary` to indicate whether we need to call `maybeUpdate` or not.
* Set `true` to the flag when the summary is updated.
* Change `maybeUpdate` to return whether it has actually updated or not.
* set `false` to the flag when the `maybeUpdate` returns `true`.

Reply:
Why are we special casing speculative execution summary compared to other updates (to stage, executor, etc).
With the change below (`maybeUpdate` only when task start is for speculative task) we should be fine ?

Reply:
> With the change below (maybeUpdate only when task start is for speculative task) we should be fine ?
> With the change below (maybeUpdate only when task start is for speculative task) we should be fine ?

 `maybeUpdate` doesn't always write to the KVStore. So, if writing to the KVStore happens only when a speculative task starts, and the speculation summary of a speculative task is not written by `maybeUpdate`, the next chance the speculation summary can be written is the time another speculative task starts

Reply:
Wont this not be the same for any of the other kv store updates in this method as well ? (just that the frequency is slightly higher than for speculative tasks).
Having said that, if that is a concern, replacing `maybeUpdate` with `update` - when done within `event.taskInfo.speculative` should be fine - it is a fairly infrequent occurrence anyay !

Reply:
@thejdeep Are you satisfied with the idea @mridulm suggests, that `maybeUpdate` happens only if a task is speculative  ? You concerned about the similar idea.
https://github.com/apache/spark/pull/34607#discussion_r750963340

Another idea that changing `maybeUpdate` to `update` would one option.
I think it's inefficient for the HistoryServer but I'm OK with this idea if a stage should not start so many speculative tasks in the real-world use cases.

Reply:
@sarutak Agreed, we can just perform an update when we find a speculative task.